Parallel mod loading causes CMEs which result in create's creative tab missing entries

Has been fixed in 0.5.1.h.

### Latest Info:
The issue has been found and a fix has been implemented, no more debugging is needed and thank you to everyone who helped in finding the issue. https://github.com/Creators-of-Create/Create/issues/6222#issuecomment-2082631617

A patch with a fix is not out yet, once it is out this issue will be closed.

@IThundxr Found a new race condition(?):

```
[10Jun2024 08:01:22.411] [modloading-worker-0/ERROR] [net.minecraftforge.fml.javafmlmod.FMLModContainer/LOADING]: Failed to create mod instance. ModID: create, class com.simibubi.create.Create
java.lang.reflect.InvocationTargetException: null
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:480) ~[?:?]
	at net.minecraftforge.fml.javafmlmod.FMLModContainer.constructMod(FMLModContainer.java:68) ~[javafmllanguage-1.20.1-47.2.0.jar%23716!/:?]
	at net.minecraftforge.fml.ModContainer.lambda$buildTransitionHandler$10(ModContainer.java:123) ~[fmlcore-1.20.1-47.2.0.jar%23715!/:?]
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
	at java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) ~[?:?]
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182) ~[?:?]
	at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[?:?]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622) ~[?:?]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165) ~[?:?]
Caused by: java.util.ConcurrentModificationException
	at java.util.HashMap.forEach(HashMap.java:1424) ~[?:?]
	at com.simibubi.create.infrastructure.config.CStress.registerAll(CStress.java:28) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.foundation.config.ConfigBase.lambda$nested$0(ConfigBase.java:83) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.foundation.config.ConfigBase$CValue.lambda$new$0(ConfigBase.java:103) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.foundation.config.ConfigBase$CValue.register(ConfigBase.java:121) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.foundation.config.ConfigBase.registerAll(ConfigBase.java:26) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.foundation.config.ConfigBase.lambda$nested$0(ConfigBase.java:83) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.foundation.config.ConfigBase$CValue.lambda$new$0(ConfigBase.java:103) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.foundation.config.ConfigBase$CValue.register(ConfigBase.java:121) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.foundation.config.ConfigBase.registerAll(ConfigBase.java:26) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.infrastructure.config.AllConfigs.lambda$register$0(AllConfigs.java:48) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at net.minecraftforge.common.ForgeConfigSpec$Builder.configure(ForgeConfigSpec.java:609) ~[forge-1.20.1-47.2.0-universal.jar%23719!/:?]
	at com.simibubi.create.infrastructure.config.AllConfigs.register(AllConfigs.java:46) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.infrastructure.config.AllConfigs.register(AllConfigs.java:61) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.Create.onCtor(Create.java:124) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	at com.simibubi.create.Create.<init>(Create.java:93) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
	... 14 more
```

Looks like the config map needs to be concurrent too.

> @IThundxr Found a new race condition(?):
> 
> 
> 
> ```
> 
> [10Jun2024 08:01:22.411] [modloading-worker-0/ERROR] [net.minecraftforge.fml.javafmlmod.FMLModContainer/LOADING]: Failed to create mod instance. ModID: create, class com.simibubi.create.Create
> 
> java.lang.reflect.InvocationTargetException: null
> 
> 	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
> 
> 	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77) ~[?:?]
> 
> 	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
> 
> 	at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[?:?]
> 
> 	at java.lang.reflect.Constructor.newInstance(Constructor.java:480) ~[?:?]
> 
> 	at net.minecraftforge.fml.javafmlmod.FMLModContainer.constructMod(FMLModContainer.java:68) ~[javafmllanguage-1.20.1-47.2.0.jar%23716!/:?]
> 
> 	at net.minecraftforge.fml.ModContainer.lambda$buildTransitionHandler$10(ModContainer.java:123) ~[fmlcore-1.20.1-47.2.0.jar%23715!/:?]
> 
> 	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
> 
> 	at java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796) ~[?:?]
> 
> 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) ~[?:?]
> 
> 	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182) ~[?:?]
> 
> 	at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[?:?]
> 
> 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622) ~[?:?]
> 
> 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165) ~[?:?]
> 
> Caused by: java.util.ConcurrentModificationException
> 
> 	at java.util.HashMap.forEach(HashMap.java:1424) ~[?:?]
> 
> 	at com.simibubi.create.infrastructure.config.CStress.registerAll(CStress.java:28) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.foundation.config.ConfigBase.lambda$nested$0(ConfigBase.java:83) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.foundation.config.ConfigBase$CValue.lambda$new$0(ConfigBase.java:103) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.foundation.config.ConfigBase$CValue.register(ConfigBase.java:121) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.foundation.config.ConfigBase.registerAll(ConfigBase.java:26) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.foundation.config.ConfigBase.lambda$nested$0(ConfigBase.java:83) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.foundation.config.ConfigBase$CValue.lambda$new$0(ConfigBase.java:103) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.foundation.config.ConfigBase$CValue.register(ConfigBase.java:121) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.foundation.config.ConfigBase.registerAll(ConfigBase.java:26) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.infrastructure.config.AllConfigs.lambda$register$0(AllConfigs.java:48) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at net.minecraftforge.common.ForgeConfigSpec$Builder.configure(ForgeConfigSpec.java:609) ~[forge-1.20.1-47.2.0-universal.jar%23719!/:?]
> 
> 	at com.simibubi.create.infrastructure.config.AllConfigs.register(AllConfigs.java:46) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.infrastructure.config.AllConfigs.register(AllConfigs.java:61) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.Create.onCtor(Create.java:124) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	at com.simibubi.create.Create.<init>(Create.java:93) ~[create-1.20.1-0.5.1.f.jar%23497!/:0.5.1.f]
> 
> 	... 14 more
> 
> ```
> 
> 
> 
> Looks like the config map needs to be concurrent too.

could you make a new issue with reproduction steps? thanks

--- a/src/main/java/com/simibubi/create/foundation/data/CreateRegistrate.java
+++ b/src/main/java/com/simibubi/create/foundation/data/CreateRegistrate.java
@@ -2,8 +2,9 @@ package com.simibubi.create.foundation.data;
 
 import static com.simibubi.create.foundation.data.TagGen.pickaxeOnly;
 
+import java.util.Collections;
+import java.util.IdentityHashMap;
 import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 import java.util.function.Function;
@@ -52,7 +53,7 @@ import net.minecraftforge.fml.DistExecutor;
 import net.minecraftforge.registries.RegistryObject;
 
 public class CreateRegistrate extends AbstractRegistrate<CreateRegistrate> {
-	private static final Map<RegistryEntry<?>, RegistryObject<CreativeModeTab>> TAB_LOOKUP = new ConcurrentHashMap<>();
+	private static final Map<RegistryEntry<?>, RegistryObject<CreativeModeTab>> TAB_LOOKUP = Collections.synchronizedMap(new IdentityHashMap<>());
 
 	@Nullable
 	protected Function<Item, TooltipModifier> currentTooltipModifierFactory;

--- a/src/main/java/com/simibubi/create/foundation/data/CreateRegistrate.java
+++ b/src/main/java/com/simibubi/create/foundation/data/CreateRegistrate.java
@@ -2,6 +2,7 @@ package com.simibubi.create.foundation.data;
 
 import static com.simibubi.create.foundation.data.TagGen.pickaxeOnly;
 
+import java.util.Collections;
 import java.util.IdentityHashMap;
 import java.util.Map;
 import java.util.function.BiConsumer;
@@ -52,7 +53,7 @@ import net.minecraftforge.fml.DistExecutor;
 import net.minecraftforge.registries.RegistryObject;
 
 public class CreateRegistrate extends AbstractRegistrate<CreateRegistrate> {
-	private static final Map<RegistryEntry<?>, RegistryObject<CreativeModeTab>> TAB_LOOKUP = new IdentityHashMap<>();
+	private static final Map<RegistryEntry<?>, RegistryObject<CreativeModeTab>> TAB_LOOKUP = Collections.synchronizedMap(new IdentityHashMap<>());
 
 	@Nullable
 	protected Function<Item, TooltipModifier> currentTooltipModifierFactory;

--- a/src/main/java/com/simibubi/create/foundation/data/CreateRegistrate.java
+++ b/src/main/java/com/simibubi/create/foundation/data/CreateRegistrate.java
@@ -2,9 +2,8 @@ package com.simibubi.create.foundation.data;
 
 import static com.simibubi.create.foundation.data.TagGen.pickaxeOnly;
 
-import java.util.Collections;
-import java.util.IdentityHashMap;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 import java.util.function.Function;
@@ -53,7 +52,7 @@ import net.minecraftforge.fml.DistExecutor;
 import net.minecraftforge.registries.RegistryObject;
 
 public class CreateRegistrate extends AbstractRegistrate<CreateRegistrate> {
-	private static final Map<RegistryEntry<?>, RegistryObject<CreativeModeTab>> TAB_LOOKUP = Collections.synchronizedMap(new IdentityHashMap<>());
+	private static final Map<RegistryEntry<?>, RegistryObject<CreativeModeTab>> TAB_LOOKUP = new ConcurrentHashMap<>();
 
 	@Nullable
 	protected Function<Item, TooltipModifier> currentTooltipModifierFactory;

